### ISSUE TITLE
Закончить спецификацию структурной разметки ЦСЯ Markdown.

### ISSUE THREAD

**Issue by 0xGeorgii** [slavonic/tooling-todo#4]
Из обсуждений следует что структурная разметка должна быть открытой (легко расширяемой). 

Единицей разметки объявлем блок (параграф).

**Comment by typiconman** [comment]
Спецификация должна учесть проблему хранения Библии. Это пока не учтено, ни в Markdown, ни в XML. Сейчас Библия хранится вот так:
https://github.com/typiconman/ponomar/blob/master/Ponomar/languages/cu/bible/elis/Acts.text

**Comment by pgmmpk** [comment]
Новое:
1. Нумерация стихов
2. Метки зачал

Может как-то вот так:

Было:
```
#1
1| **(Зача́ло 1)** Пе́рвое ᲂу҆́бѡ сло́во сотвори́хъ ѡ҆ всѣ́хъ, ѽ, ѳео́фїле, ꙗ҆̀же нача́тъ і҆и҃съ твори́ти же и҆ ᲂу҆чи́ти 
2| да́же до днѐ, во́ньже, заповѣ́давъ а҆пⷭ҇лѡмъ дх҃омъ ст҃ы́мъ, и҆̀хже и҆збра̀, вознесе́сѧ: 
```

Стало (Markdown):
```
Глава а

а. **(Зача́ло 1)** Пе́рвое ᲂу҆́бѡ сло́во сотвори́хъ ѡ҆ всѣ́хъ, ѽ, ѳео́фїле, ꙗ҆̀же нача́тъ і҆и҃съ твори́ти же и҆ ᲂу҆чи́ти 

в. да́же до днѐ, во́ньже, заповѣ́давъ а҆пⷭ҇лѡмъ дх҃омъ ст҃ы́мъ, и҆̀хже и҆збра̀, вознесе́сѧ: 
```

**Comment by typiconman** [comment]
Гм. Думаю, что лучше все таки арабские цифры. Потому что в конце концов, мы же можем хранить в этом формате тексты не только на ЦСЯ.

**Comment by 0xGeorgii** [comment]
why can't we store it in the HTML or XML format? it's much more suitable for further machine handling. Or even leave it as it is.

**Comment by pgmmpk** [comment]
@typiconman 

Для ЦСЯ используем буквенную цифирь (как в книге). Для гражданицы или латинского - арабские цифры (опять же - как в книге). Иначе придется задачу отображения (перевод в буквенную цифирь или что там в книге) решать средствами Markdown (я предполагаю что для ЦСЯ показывать на экране хотим цифирью).

Вот чего не хватает в предлагаемой разметке - это анкоров для навигации. Например надо перейти на стих 2. Это можно решить средствами структурной разметки - пометить каждый блок-стих

**Comment by pgmmpk** [comment]
@GeorgePlotnikov проблема не с машиной а с людьми. Человеки не должны редактировать XML. Тогда встает проблема инструментария.

Markdown - это, собственно, версия такого инструментария.

**Comment by typiconman** [comment]
Пометить каждый блок-стих? Что Вы имеете ввиду, @pgmmpk?

**Comment by pgmmpk** [comment]
> Пометить каждый блок-стих? Что Вы имеете ввиду, @pgmmpk?

@typiconman ну, мы же планируем структурную аннотацию в виде произвольных ключей и значений (a-la RDFa). Тогда можно вот так:

```
Глава а

{{verse=1}}
а. **(Зача́ло 1)** Пе́рвое ᲂу҆́бѡ сло́во сотвори́хъ ѡ҆ всѣ́хъ, ѽ, ѳео́фїле, ꙗ҆̀же нача́тъ і҆и҃съ твори́ти же и҆ ᲂу҆чи́ти 

{{verse=2}}
в. да́же до днѐ, во́ньже, заповѣ́давъ а҆пⷭ҇лѡмъ дх҃омъ ст҃ы́мъ, и҆̀хже и҆збра̀, вознесе́сѧ: 
```

**Comment by 0xGeorgii** [comment]
> > Пометить каждый блок-стих? Что Вы имеете ввиду, @pgmmpk?
> 
> @typiconman ну, мы же планируем структурную аннотацию в виде произвольных ключей и значение (a-la RDFa). Тогда можно вот так:
> 
> ```
> Глава а
> 
> {{verse=1}}
> а. **(Зача́ло 1)** Пе́рвое ᲂу҆́бѡ сло́во сотвори́хъ ѡ҆ всѣ́хъ, ѽ, ѳео́фїле, ꙗ҆̀же нача́тъ і҆и҃съ твори́ти же и҆ ᲂу҆чи́ти 
> 
> {{verse=2}}
> в. да́же до днѐ, во́ньже, заповѣ́давъ а҆пⷭ҇лѡмъ дх҃омъ ст҃ы́мъ, и҆̀хже и҆збра̀, вознесе́сѧ: 
> ```

headers in the MD automatically converts to the header+anchor

**Comment by pgmmpk** [comment]
Чтобы уточнить. Дело не в XML против Markdown. Разница между ними лишь синтаксическая. Реально содержательная работа - это описать необходимую семантику. Будь это XML схема или спецификации ЦСЯ Markdown.

Пока минимальные требования к разметке вроде таковы:

Структурно:
1. заголовки 
2. параграфы
3. произвольная мета-информация на уровне файла и параграфа (key=>value). В общем случае мета не отображается. Используется для навигации, композиции, поиска
4. выносная (на полях) нумерация
5. начало новой страницы (для координации с картинками факсимиле)

Inline:
1. Киноварь
2. Красная буква
3. Буквица
4. Разрядка
5. Переключение языка (cu, ru, gr, en, bu)
6. Сноски
7. Метка зачала (начало - конец)

(надеюсь ничего не пропустил)

В этом свете кажется что формат Библии почти ничего нового не добавляет. Главы отображаются на заголовки (которые Markdown автоматически снабжает анкорами, как правильно заметил @GeorgePlotnikov ). Стихи становятся параграфами. Новый inline элемент - это метка зачала. 

Что я пропустил?

**Comment by 0xGeorgii** [comment]
@pgmmpk cross-references maybe?

**Comment by pgmmpk** [comment]
> @pgmmpk cross-references maybe?

Вопрос интересный. На уровне `<a href>` это уже есть в MD и HTML. Однако рассмотрим вот это:
```
[Мф.1-5]
```

Можно было бы привязать к конкретной библии и захардкодить что-нибудь вроде `http://bibliya.ru/mathew#1-5`. Но это получается грубо. Ссылаются-то на стих Матфея вообще, а не на какое-то конкретное печатное издание.

Правильнее (наверное) прописать что-то вроде `{{book=mathew, chapter=1, verse=5}}` и поручить приложению решать куда отослать кликающего.

 Соответственно в том же корпусе книг может быто одна или несколько книг, которые объявляют себя "евангелием от Матфея" и нужный параграф там должен быть помечен `{{verse=5}}`. Тогда приложение сможет разрешить ссылку. 

А более продвинутое приложение сможет проиндексировать ещё несколько сайтов и предложить кликающему выбор куда пойти...

Короче, концепция ЦСЯ ссылок требует разработки

**Comment by typiconman** [comment]
Не кажется ли Вам, @pgmmpk, что синтаксис

`{{verse=2}}
в. ...`

является дупликацией? Одно можно сгенирировать из другого в run time.

И потом, строго говоря, тогда должно быть `в҃` (т.е. с титло).

**Comment by pgmmpk** [comment]
Да, Александр, есть дупликация тут. Извините за пропущенное титло - оно имелось в виду быть.

Гм. Наверное можно заставить MD реагировать на ключ `verse` и автоматически добавлять цифирь к последующему параграфу (если документ lang=cu => буквенную, else арабскую).

А как удобнее пользователю? Держать в голове это правило или полагаться на более очевидное "что пишем то и видим"?

**Comment by typiconman** [comment]
Кажется это уже вопрос к инструментарию. Скажем у нас какое-то приложение. Пользователь запрашивает Мф. 1:5. Приложение открывает файл с Мф, находит главу 1, там ноходит verse=5 и считывает этот стих. Дальше оно конвертирует текст для отображения пользователю в HTML (или во что-то еще). На этом этапе легко сконвертировать 5 -> є҃, средствами того же Lingua::CU  и т.п.
Или не заморачиваться, и просто хранить сразу номера стихов? Но в некоторых случаях, пользователь может запросить текст без номеров стихов. Нужно ему дать опцию не отображать нумерацию. Как сейчас сделано здесь:
https://ponomar.net/cgi-bin/bible.cgi

**Comment by pgmmpk** [comment]
> Нужно ему дать опцию не отображать нумерацию

Это решает вопрос - нумерацию надо убрать из текста и генерировать динамически из аннотации `{{verse}}`

**Comment by 0xGeorgii** [comment]
100% this process *must* be automated

**Comment by typiconman** [comment]
Еще два вопроса:
1. Новая строка после {{verse}} обязательна? Т.е. `{{verse=1}} Lorum ipsum` или `{{verse=1}}
Lorum ipsum`?

2. Нумерация глав необходима? `{{chapter=1}}`? При этом нужно писать `{{chapter=1}} Глава а҃`?

**Comment by pgmmpk** [comment]
> Нумерация глав необходима? {{chapter=1}}? При этом нужно писать {{chapter=1}} Глава а҃?

Да, думаю что в общем случае необходима. Если есть необходимость прятать главы, то можно поступить также как и с {{verse}} - поручить генерацию глав движку.

> Новая строка после {{verse}} обязательна? Т.е. {{verse=1}} Lorum ipsum или {{verse=1}} Lorum ipsum?

Нет необязательна. Более того, можно и так: `Lorum ipsum {{verse=1}}` и даже так `Lorum {{verse=1}} ipsum`. Важно отсутствие пустой строки между аннотацией и блоком текста.

**Comment by pgmmpk** [comment]
Меня терзают смутные сомнения по поводу HTML/XML генерации из `{{verse=x}}`. Собственно хочется просто реализовать структурные аннотации как атрибуты блока, не разбирая их семантики. Примерно так:

Вход:
```
{{tone=8 day=mon}}
Lorem ipsum
```

Выход:
```
<p tone="8" day="mon">Lorem ipsum</p>
```

То есть MD движок просто переносит атрибуты в блок.

Если мы так делаем, то таксономия атрибутов аннотации может быть совершенно открытая
и распределенная (похоже на RDFa). Это здорово усиливает возможности стандарта.

Однако как быть с `{{verse=1}}`?

Может вытеснить генерацию цифири за пределы HTML? В JavaScript?

Или предлагать желающим создавать свои расширения ЦСЯ MD?

Давайте подумаем вместе.
